File: 1036-rotting-oranges/rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        int mins=0;
        int fresh=0;
        int rowLen=grid.length;
        int colLen=grid[0].length;

        //find a rotten orange
        Queue<int[]> queue = new LinkedList<>();
        for(int i=0;i<rowLen;i++) {
            for(int j=0;j<colLen;j++) {
                if(grid[i][j]==2) {
                    queue.add(new int[]{i,j});
                }
                if(grid[i][j]==1) {
                    fresh++;
                }
            }
        }

        if (fresh == 0) return 0;

        while(!queue.isEmpty() && fresh > 0) {
            int queueSize = queue.size();

            for(int m=0;m<queueSize;m++) {
                int[] elem = queue.poll();

                if(elem[0]-1>=0 && grid[elem[0]-1][elem[1]]==1) {
                    grid[elem[0]-1][elem[1]]=2;
                    fresh--;
                    queue.add(new int[]{elem[0]-1,elem[1]});
                }
                if(elem[0]+1 < rowLen && grid[elem[0]+1][elem[1]]==1) {
                    grid[elem[0]+1][elem[1]]=2;
                    fresh--;
                    queue.add(new int[]{elem[0]+1,elem[1]});
                }
                if(elem[1]-1>=0 && grid[elem[0]][elem[1]-1]==1) {
                    grid[elem[0]][elem[1]-1]=2;
                    fresh--;
                    queue.add(new int[]{elem[0],elem[1]-1});
                }
                if(elem[1]+1 < colLen && grid[elem[0]][elem[1]+1]==1) {
                    grid[elem[0]][elem[1]+1]=2;
                    fresh--;
                    queue.add(new int[]{elem[0],elem[1]+1});
                }
            }
            mins++;
        }
        return fresh == 0 ? mins : -1;
    }
}

File: 1036-rotting-oranges/rotting-oranges.cpp
class Solution {
public:
    struct Position {
        int row;
        int col;
        int mins;
    };
    int orangesRotting(vector<vector<int>>& grid) {
        int row_size=grid.size();
        int col_size=grid[0].size();

        Position queue[10000];
        int front=0;
        int rear=0;

        bool hasFresh=false;
        //first pushes
        for(int i=0;i<row_size;i++) {
            for(int j=0;j<col_size;j++) {
                if(grid[i][j]==2) {
                    queue[rear].row=i;
                    queue[rear].col=j;
                    queue[rear].mins=0;
                    rear++;
                }

                if(grid[i][j]==1) {
                    hasFresh=true;
                }
            }
        }

        //check if there is no fresh orange
        if(!hasFresh) {
            return 0;
        }

        while(front<rear) {
            //pop
            Position current = queue[front];
            front++;

            //check if every orange is rotten
            bool allRotten=true;
            for(int i=0;i<row_size;i++) {
                for(int j=0;j<col_size;j++) {
                    if(grid[i][j]==1) {
                        allRotten=false;
                        break;
                    }
                }
            }

            if(allRotten && front==rear) {
                return current.mins;
            }
            

            //top
            if(current.row>0 && grid[current.row-1][current.col]==1) {
                //mark rotten
                grid[current.row-1][current.col]=2;

                //push
                queue[rear].row=current.row-1;
                queue[rear].col=current.col;
                queue[rear].mins=current.mins+1;
                rear++;
            }

            //bottom
            if(current.row<row_size-1 && grid[current.row+1][current.col]==1) {
                //mark rotten
                grid[current.row+1][current.col]=2;

                //push
                queue[rear].row=current.row+1;
                queue[rear].col=current.col;
                queue[rear].mins=current.mins+1;
                rear++;
            }

            //left
            if(current.col>0 && grid[current.row][current.col-1]==1) {
                //mark rotten
                grid[current.row][current.col-1]=2;

                //push
                queue[rear].row=current.row;
                queue[rear].col=current.col-1;
                queue[rear].mins=current.mins+1;
                rear++;
            }

            //right
            if(current.col<col_size-1 && grid[current.row][current.col+1]==1) {
                //mark rotten
                grid[current.row][current.col+1]=2;

                //push
                queue[rear].row=current.row;
                queue[rear].col=current.col+1;
                queue[rear].mins=current.mins+1;
                rear++;
            }

        }

        return -1;
    }
};